design/accepted/kv_auth.md

```diff
@@ -0,0 +1,115 @@
+- Deleted by `Username`.
+- Listed by `Username`.
+
+### Group
+
+- Looked up, deleted and listed by `DisplayName`.
+
+### Policies
+
+- Looked up, deleted and listed by `DisplayName`.
+
+### Credentials
+
+- Has a foreign id reference to `User.ID`.
+    - The `User.ID` is used for looking up the `User.Username`.
+    - Almost all actions are in the context of `User.Username`, i.e. it is always passed.
+- The only action without a `User.Username` passed is credentials lookup by `AccessKeyID`.
+
+
+## Entities Relationship
+
+### Group <-> User
+
+- Add/Remove `User` to `Group` by `User.Username` & `Group.DisplayName`. 
+- List all the `Group`s for a `User` by `User.Username`.
+- List all the `Users`s in a `Group` by `Group.DisplayName`.
+
+### Policy <-> User
+
+- Attach/Detach `Policy` to `User` by `User.Username` & `Policy.DisplayName`.
+- List all `Policies` for a `User` by `User.Username`.
+- List all effective `Policies` for a `User` by `User.Username`, this includes all `Policies` for `Group`s is a member of.
+
+### Policy <-> Group
+
+- Attach/Detach `Policy` to `Group` by `Group.DisplayName` & `Policy.DisplayName`.
+- List all `Policies` for a `Group` by `Group.DisplayName`.
+    - `Group` cannot be a member of another `Group`, therefore all effective policies are attached to it directly.
+
+## Representation in the KV world
+
+- All keys are prefixed by the reserved package prefix `auth/`.
+  
+- `User`s key will be in the form of `users/<UserName>`.
+- `Group`s key will be in the form of `groups/<DisplayName>`.
+- `Policies`s key will be in the form of `policies/<DisplayName>`.
+- `Credentials`s key will be in the form of `users/<UserName>/credentials/<AccessKeyID>`.
+
+### Handling Secondary Indexes
+
+Below are 2 options that vary in efficiency and complexity of the implementation.
+ 
+#### Store just the minimum indexes
+
+Keep only the minimum that is required to represent the relationship of the entities:
+
+- A `User` membership of a `Group` under `groups/<DisplayName>/users/<Username>`.
+- A `Policy` attached to a `User` under `users/<Username>/policies/<DisplayName>`.
+- A `Policy` attached to a `Group` under `groups/<Displayname>/policies/<DisplayName>`.
+
+Any deletion of an entity would first remove all its secondary indexes, then the entity itself.
+For example, deleting a `Policy` would have to list all `User`s & `Group`'s `Policies` and delete any attachment if existed.
+Only then it should delete the `Policy` entity from the store.
+
+Listing by anything other than the key will require to list all the entities which are relevant.
+Some examples:
+
+1. Finding a `User` by `AccessKeyID` will require to list all the users.
+1. Listing the `User`'s effective policies will require to list all entities under `groups/`.
+
+The amount of entities in the Auth world isn't big (<10k) and cached in the server, it's unlikely it will incur a 
+notable performance degradation.
+
+#### Store all secondary indexes
```

Comment:
I am more worried about duplication and the risk of inconsistency than performance itself. And since your assumption is that performance in this data scale is not an issue - I think we should make sure not to duplicate data unless it is absolutely unavoidable

Reply:
+1 to @N-o-Z 
due to the quantitive nature of this feature (small amounts in IAM), I would prefer consistency over performance, especially with the cache.

Reply:
I agree, that's why I preferred option 1.